src/core/tsi/s2a/handshaker/s2a_tsi_handshaker.cc

```diff
@@ -16,34 +16,327 @@
+#include <grpc/support/log.h>
+#include <grpc/support/string_util.h>
+#include <grpc/support/sync.h>
+
+#include "src/core/lib/slice/slice_internal.h"
+#include "src/core/lib/surface/channel.h"
+#include "src/core/tsi/s2a/handshaker/s2a_handshaker_client.h"
+#include "src/core/tsi/s2a/s2a_constants.h"
+
+namespace grpc_core {
 namespace experimental {
 
+/** The main struct for the S2A TSI handshaker. **/
+struct s2a_tsi_handshaker {
+  tsi_handshaker base;
+  grpc_slice target_name;
+  bool is_client;
+  bool has_sent_start_message;
+  bool has_created_handshaker_client;
+  char* handshaker_service_url;
+  grpc_pollset_set* interested_parties;
+  grpc_s2a_credentials_options* options;
+  s2a_handshaker_client_vtable* client_vtable_for_testing;
+  grpc_channel* channel;
+  gpr_mu mu;
+  s2a_handshaker_client* client;
+  bool shutdown;
+};
+
+/** The main struct for the S2A TSI handshaker result. **/
+typedef struct s2a_tsi_handshaker_result {
+  tsi_handshaker_result base;
+  uint16_t tls_version;
+  uint16_t tls_ciphersuite;
+  uint8_t* in_traffic_secret;
+  uint8_t* out_traffic_secret;
+  size_t traffic_secret_size;
+  char* spiffe_id;
+  char* hostname;
+  unsigned char* unused_bytes;
+  size_t unused_bytes_size;
+  bool is_client;
+} s2a_tsi_handshaker_result;
+
+static tsi_result handshaker_next(
+    tsi_handshaker* self, const unsigned char* received_bytes,
+    size_t received_bytes_size, const unsigned char** /** bytes_to_send **/,
+    size_t* /** bytes_to_send_size **/, tsi_handshaker_result** /** result **/,
+    tsi_handshaker_on_next_done_cb cb, void* user_data) {
+  // TODO(mattstev): implement.
+  return TSI_UNIMPLEMENTED;
+}
+
+static void handshaker_shutdown(tsi_handshaker* self) {
+  GPR_ASSERT(self != nullptr);
+  s2a_tsi_handshaker* handshaker = reinterpret_cast<s2a_tsi_handshaker*>(self);
+  grpc_core::MutexLock lock(&(handshaker->mu));
+  if (handshaker->shutdown) {
+    return;
+  }
+  if (handshaker->client != nullptr) {
+    s2a_handshaker_client_shutdown(handshaker->client);
+  }
+  handshaker->shutdown = true;
+}
+
+static void handshaker_destroy(tsi_handshaker* self) {
+  if (self == nullptr) {
+    return;
+  }
+  s2a_tsi_handshaker* handshaker = reinterpret_cast<s2a_tsi_handshaker*>(self);
+  s2a_handshaker_client_destroy(handshaker->client);
+  grpc_slice_unref_internal(handshaker->target_name);
+  // TODO(mattstev): this API is exposed in a PR that is not yet merged.
+  // grpc_s2a_credentials_options_destroy(handshaker->options);
+  if (handshaker->channel != nullptr) {
+    grpc_channel_destroy_internal(handshaker->channel);
+  }
+  gpr_free(handshaker->handshaker_service_url);
+  gpr_mu_destroy(&(handshaker->mu));
+  gpr_free(handshaker);
+}
+
+static const tsi_handshaker_vtable handshaker_vtable = {
+    nullptr,         nullptr,
+    nullptr,         nullptr,
+    nullptr,         handshaker_destroy,
+    handshaker_next, handshaker_shutdown};
+
 tsi_result s2a_tsi_handshaker_create(
```

Comment:
We probably do not need `handshaker_service_url` parameter since the information is already a part of `grpc_s2a_credentials_options`.

Reply:
Agreed. Will remove in the next commit.